Small fixes to improve TensorIterator overhead for the common case of inputs and outputs of the same type

ngimel opened the pull request:
Summary:
1) Short-circuits computing common type and type promotion logic for the common case of operands and result of the same type
2) Improves performance of checking memory overlap by returning MemoryOverlap::FULL if tensors are the same, skips the call
from TensorIterator when tensors are the same
3) Changes the default size of DimVector from 5 to 6, thus allowing it not to be resized for a common case of binary operation. `strides`
DimVector is forced to have at least 2*num_tensors elements, which for an operation with 2 inputs and one output is 6
4) If `offset` is 0 (common non-broadcasting case), don't fill `strides` vector with 0-s, because all the values will be subsequently written to.

These changes combined improve the overhead from 1.02 us to .74 us for a simple in-place operation.

Test Plan: should be covered by existing tests

Differential Revision: D17784532

ngimel commented:
cc @nairbv for the TODO item https://github.com/pytorch/pytorch/pull/27457/files#diff-b47a50873394e38a005b4c1acd151957R116

colesbury commented on aten/src/ATen/core/DimVector.h:
```diff
@@ -5,7 +5,7 @@
 namespace at {
 
-constexpr size_t kDimVectorStaticSize = 5;
+constexpr size_t kDimVectorStaticSize = 6;
```
This is for get_strides(), right? That function is weird because it stores `collapsed_dims x n_tensors` dimensions whereas DimVector is "normally" just the dims of a single Tensor.

It may just be better to change that function to use `SmallVector<int64_t, 6>` (larger than 6 fine too;  define a new type alias if you like). 

I don't think TensorImpl uses DimVector directly anymore, but it also uses 5 dims. It would be nice to keep this consistent with that.

resistor commented on aten/src/ATen/native/TensorIterator.cpp:
```diff
@@ -702,7 +730,9 @@ void TensorIterator::check_mem_overlaps() {
     for (int j = num_outputs_; j < ntensors(); j++) {
       const auto& input = operands_[j].tensor;
-      assert_no_partial_overlap(output, input);
+      if (!output.is_same(input)) {
```
Should this check be sunk into `assert_no_partial_overlap`?

ngimel replied:
I already added a short-circuit in get_overlap_status, but it was behind .contiguous so was slower, now I pulled it up there and removed from here, seems to be ok.

Commit:
Small fixes to improve TensorIterator overhead for the common case of inputs and outputs of the same type (#27457)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/27457

1) Short-circuits computing common type and type promotion logic for the common case of operands and result of the same type
2) Improves performance of checking memory overlap by returning MemoryOverlap::FULL if tensors are the same, skips the call
from TensorIterator when tensors are the same
3) Changes the default size of DimVector from 5 to 6, thus allowing it not to be resized for a common case of binary operation. `strides`
DimVector is forced to have at least 2*num_tensors elements, which for an operation with 2 inputs and one output is 6
4) If `offset` is 0 (common non-broadcasting case), don't fill `strides` vector with 0-s, because all the values will be subsequently written to.

These changes combined improve the overhead from 1.02 us to .74 us for a simple in-place operation.

Test Plan: should be covered by existing tests

Differential Revision: D17784532

fbshipit-source-id: 5e523e1892cfe4b7aee0d90cc8e820409c309e9f

VitalyFedyunin commented on aten/src/ATen/native/TensorIterator.h:
```diff
@@ -134,6 +135,7 @@ enum class CommonDTypeStrategy : uint8_t {
 struct CAFFE2_API TensorIterator {
   using DimMask = std::bitset<64>;
   using PtrVector = SmallVector<char*, 4>;
+  using StrideVector = SmallVector<int64_t, 6>;
```
super nit: maybe move this magical numbers into constant with good name
super super nit: make Ptr vector same size as StrideVector
